### PR TITLE
rgw: add support for the prefix parameter in account listing of Swift API

### PR DESCRIPTION
**Pull request by rzarzynski**
Fixes: http://tracker.ceph.com/issues/17931

### CODE REVIEW

**Comment by rzarzynski** [comment]
No regression in Tempest. S3 isn't affected by the PR.

**Comment by rzarzynski** [comment]
@yehudasa: refactored the dependency on `g_conf` by switching to `s->cct->_conf` in the whole  `RGWListBuckets_ObjStore_SWIFT` class.

Tempest is still fine with this PR.

**Comment by yehudasa** [comment]
@rzarzynski I still see use of g_conf there

**Comment by rzarzynski** [comment]
@yehudasa: which places do you mean? I pushed the `g_conf` refactor as a separated commit https://github.com/ceph/ceph/pull/12047/commits/1d7e94b2afbadb661214480f3edb21d4462bd768 to this PR.

**Comment by rzarzynski** [comment]
@yehudasa: have you meant the occurrences of `g_conf` in `rgw_rest_swift.cc` other than those in `RGWListBuckets_ObjStore_SWIFT`?

**Comment by yehudasa** [comment]
@rzarzynski no, it's ok now, not sure what happened before.

File: src/rgw/rgw_rest_swift.cc [hunk 1]
```diff
@@ -170,22 +171,28 @@ void RGWListBuckets_ObjStore_SWIFT::send_response_begin(bool has_buckets)
...
+
     s->formatter->open_object_section("container");
     s->formatter->dump_string("name", obj.bucket.name);
     if (need_stats) {
       s->formatter->dump_int("count", obj.count);
       s->formatter->dump_int("bytes", obj.size);
     }
     s->formatter->close_section();
-    if (!g_conf->rgw_swift_enforce_content_length) {
+    if (! g_conf->rgw_swift_enforce_content_length) {
```
Comment (yehudasa) [inline src/rgw/rgw_rest_swift.cc]: @rzarzynski actually, see this one ^^^

Comment (yehudasa) [inline src/rgw/rgw_rest_swift.cc]: @rzarzynski oh, nevermind

File: src/rgw/rgw_rest_swift.cc [hunk 2]
```diff
@@ -177,7 +177,7 @@ void RGWListBuckets_ObjStore_SWIFT::send_response_data(RGWUserBuckets& buckets)
     return;
 
   for (iter = m.begin(); iter != m.end(); ++iter) {
-    RGWBucketEnt obj = iter->second;
+    const RGWBucketEnt& obj = iter->second;
```
Comment (mattbenjamin) [inline src/rgw/rgw_rest_swift.cc]: ok

File: src/rgw/rgw_rest_swift.cc [hunk 3]
```diff
@@ -170,22 +171,28 @@ void RGWListBuckets_ObjStore_SWIFT::send_response_begin(bool has_buckets)
...
+  if (! sent_data) {
     return;
+  }
 
-  for (iter = m.begin(); iter != m.end(); ++iter) {
+  /* Take care of the prefix parameter of Swift API. There is no business
+   * in applying the filter earlier as we really need to go through all
+   * entries regardless of it (the headers like X-Account-Container-Count
+   * aren't affected by specifying prefix). */
+  const std::map<std::string, RGWBucketEnt>& m = buckets.get_buckets();
```
Comment (mattbenjamin) [inline src/rgw/rgw_rest_swift.cc]: ok

File: src/rgw/rgw_rest_swift.cc [hunk 4]
```diff
@@ -146,7 +146,7 @@ void RGWListBuckets_ObjStore_SWIFT::send_response_begin(bool has_buckets)
     set_req_state_err(s, op_ret);
   }
 
-  if (!g_conf->rgw_swift_enforce_content_length) {
+  if (! s->cct->_conf->rgw_swift_enforce_content_length) {
```
Comment (mattbenjamin) [inline src/rgw/rgw_rest_swift.cc]: ok

**Pull request merged**